server/service/schedule/schedule.go

```diff
@@ -313,3 +334,49 @@ func (s *Schedule) acquireLock() bool {
 	level.Debug(s.logger).Log("msg", "not the lock leader, skipping")
 	return false
 }
+
+func (s *Schedule) releaseLock() {
+	err := s.locker.Unlock(s.ctx, s.getLockName(), s.instanceID)
+	if err != nil {
+		level.Error(s.logger).Log("msg", "unlock failed", "err", err)
+		sentry.CaptureException(err)
+	}
+}
+
+// holdLock attempts to acquire a schedule lock. If it successfully acquires the lock, it starts a
+// goroutine that periodically extends the lock, and it returns `true` along with a
+// context.CancelFunc that will end the goroutine and release the lock. If it is unable to initially
+// acquire a lock, it returns `false, nil`. The maximum duration of the hold is two hours.
+func (s *Schedule) holdLock() (bool, context.CancelFunc) {
+	if ok := s.acquireLock(); !ok {
+		return false, nil
+	}
+
+	const MAX_HOLD = 2 * time.Hour
```

Comment:
Should this be `2 * schedInterval` instead? Also, why do we need this? What if the task does take longer than 2hrs? I mean, it's probably a big problem hah, but should we cover that case?

Reply:
If you set this as `2 * schedInterval` then you would be able to test this code path as well.

Reply:
I debated this for awhile and  may have ended up overthinking it. The deadline was intended as failsafe for a case where the cancelFunc was never called, which would probably only happen in the case of a developer oversight in the future. So maybe that is overly paternalistic and we don't need to be complicating things for developer error? The schedule interval + config interval + lock expirations is already a lot to reason about and I'm thinking now that adding another related deadline is just adding to that cognitive load.

Reply:
Just saw your comment about 2.1hr runtime above and I think that makes me even more inclined to remove this deadline altogether.

Reply:
It's not bad to have a failsafe, and it'll be very unlikely anything will take over 2hrs... Up to you if you want to leave it, but I wouldn't hardcode it if you do.

Reply:
Ok, I think I'll remove it. Like you said, it's unlikely in practice for anything to take over 2 hours so hardcoding that made sense to me. But since schedule intervals can be user-controlled, I think it is harder to confidently assess the likelihood that the 2 * schedule interval won't occur in practice.